common-components/schemas/type/money.md

```diff
@@ -0,0 +1,35 @@
+# Money
+
+The `Money` type represents an amount of money in a specified currency.
+
+## Schema
+
+A `Money` has two fields:
+
+- The `currency_code` field is a string containing a currency code.
+
+  The three-letter currency codes defined in the ISO 4217 standard are always
+  supported.
+
+  APIs may define additional currency codes that are not included in ISO 4217
+  (for example, virtual currencies or cryptocurrencies). These must start with
+  with "X-", in order to distinguish them from potential future ISO 4217 codes.
+
+  Examples:
+
+  - "USD" - ISO 4217 code for United States dollar
+  - "X-BTC" - Potential API-defined extension for Bitcoin.
+  - "X-.RBX" - Potential API-defined extension for the virtual currency Robux
+
+- The `amount` field is a field of type [`Decimal`][], representing the amount
+  of currency.
```

Comment:
This conflicts with prior art such as Agoric's [Amount](https://docs.agoric.com/glossary/#amount), which is composed of a `brand` (generalizing currency code to arbitrary distinguishable kinds) and a `value` (which is either a fungible nonnegative integer or a non-fungible unique asset). It would be unfortunate for some software to use "amount" for the composite type and other software to use it for just the component of a composite type upon which arithmetic can be performed. Is there an alternative that would work here?

Reply:
Hmm, I hadn't seen that before, interesting.

I think `value` is generally a bad field name, unless the schema itself is specifically wrapping just that one field (e.g. `google.protobuf.BoolValue`), because the whole schema represents the value.

We could go with something like `quantity` (or `count`, but that seems much worse).  But I'd rather do that because we think `quantity` is a better field name than `amount` in the API context, rather than because some crypto startup has dibs on the word "amount".

Reply:
True, but representing currency is certainly more germane to that space than it is to generic API guidelines. I'm not proposing an outright adoption of that data model, just something that avoids "amount" referring to the composite in some domains and to a field _of_ the composite in others. But even in common English, financial "amount" seems to correspond better with the former—consider e.g. "_amount due: $123.45_".

I could see using `quantity`—although it also seems to represent the composite in at least some scientific discourse, that is further removed from this purpose.

Reply:
> But even in common English, financial "amount" seems to correspond better with the former—consider e.g. "amount due: $123.45".

I think that's a good point.  Hmm.  I also think you're right that `quantity` implies a unit as well as a number.

`google.type.Money` uses two integer fields, `units` and `nanos`, in place of this field.  I don't think that really helps.

I am still pretty against `value`.  Any other ideas for what to call this field?

Reply:
Discussed in the steering meeting today:

* This field should be named "quantity".
* If/when we do a Measurement type, we should also call the corresponding field "quantity".

Reply:
@lukesneeringer https://google.aip.dev/141 specifies that "quantities" are discrete.  Any problem with using "quantity" to refer to a decimal value here?  Should AIP-141 be updated to "support" decimal values (e.g. 0.5 miles)?